### ISSUE TITLE
[Discussion] Future Research & Use Cases

### ISSUE THREAD

**Issue by MLWave** [scikit-tda/kepler-mapper#35]
# Research directions
- [ ] Multiscale mapper.
   - [ ] persistent homology of mapper - include visualization of barcode
   - [ ] choose the most stable mapper based on barcode
- [ ] Implement parameter tuning. 
- [ ] Simplicial Complex Feature extraction. 
- [x] Embeddings, Word2Vec or otherwise. 
- [x] Self-Guessing / Self-Dissimilarity. 
- [ ] Random Topological Forests. 

# Business Cases
- [x] Transactional Fraud & AML.  PII data.
- [x] Acquisition fraud.  PII data.
- [ ] Access control. Business sensitive data.
- [ ] Firewall monitoring. Business sensitive data.
- [x] Credit risk & Reject Inference. PII data.
- [ ] Customer management (churn, high spenders). PII data.

**Comment by sauln** [comment]
I suggest we have the top level comment on this issue be a list of action items or general directions with check boxes. We can then expound on each in comments below. We can rank these and continue to add to one list all of the ideas that have cropped up throughout all the issues.

**Comment by sauln** [comment]
I'm very interested in using embeddings as a lens to explore different data sets. I've been trying to explore word2vec embeddings using mapper, but haven't gotten any interesting results yet.

Do you see the embedding itself as the lens? I think we'd still need to reduce the lens down to ~3 dimensions before building the mapper.

**Comment by sauln** [comment]
More short term, I'd like to incorporate

- [ ] Multiscale mapper. 
  - Compute persistent homology and visualize barcodes.
  - Compute 'the best' mapper given a lens. Essentially do hyperparameter optimization using persistence.
- [ ] Implement parameter tuning.
  - Incorporate ideas from Oudot's new parameter tuning paper: https://arxiv.org/abs/1706.00204

**Comment by MLWave** [comment]
**Business cases**

At work (a fintech with credit card and bank account products), I use anomaly and credit score lenses to provide analysts with overviews of todays, last months, and all customers. Tooltips link to the back-end user profiles. I'm gathering feedback, to make TDA more accessible to people without knowledge of abstract algebra (but a lot of domain knowledge).

We are implementing dashboards for:
- Transactional fraud & AML. Lens [Isolation Forest, L2Norm]. Color [prior label information]. Tooltips [Links to backend profiles]
- Acquisition fraud (Identity fraud & Friends&Family fraud). Lens [Isolation Forest, L2Norm]. Color [prior label information]. Tooltips [Links to backend profiles]
- Access control. Lens [Isolation Forest, GLOSH]. Color [GLOSH outlier score]. Tooltips [Names and squads/departments of employees]
- Firewall monitoring. Lens [Isolation Forest, GLOSH]. Color [GLOSH outlier score]. Tooltips [IP address].
- Credit risk / reject inference model doctoring. Lens [Credit Score Prediction/distance to rejected population], Color [prior default information], Tooltips [Links to backend profiles]
- Customer management case studies. Lens and color [anything interesting: default, spend, churn ...], Tooltips [Cluster ID assignments].

**Comment by MLWave** [comment]
**Embeddings**

Embeddings, like those from FaceNet, GLOVE, Doc2Vec, or Word2Vec (see [FastText](https://github.com/facebookresearch/fastText), GenSim and pre-trained models from companies like [Google and Facebook](http://ahogrammer.com/2017/01/20/the-list-of-pretrained-word-embeddings/)), can be reduced down to 2 or 3 dimensions with PCA, LLE, or t-SNE.

Just works: Use the original (128 or 150 dimensional) embeddings or LSA (TFIDF2grams->SVD100d)as the inverse image to cluster on. Also works: Bag of words encode a document for the inverse image. Works better (but maybe cheating): tfidf scaling, 2-3 grams/chargrams, removing stopwords, stemming, etc. I mean... you have to represent the original data somehow right, can't feed ascii into a cluster algorithm? Or am I thinking too hard: Mapper accepts any distance matrix, as long as I properly state whether this distance satisfies triangle inequality.

This library by Facebook Research allows us to create embeddings of all the things (part of LeCun's object2vec vision): https://github.com/facebookresearch/StarSpace I think using embeddings of text documents or social graphs as lenses is an interesting venue for further research.

**Comment by MLWave** [comment]
**Self-Guessing**
- Self-Guessing introduced in A Mathematical Theory of Generalization series, D. Wolpert (1990).
10-fold Cross-validation (observing the generalization performance of a set of models on holdout data in a loop) and Stacked Generalization (using information from the generalization performance of base models as input for another model stacked on top) are primitive forms of Self-Guessing. A Cross-validation 3D mapping in combination with other lenses, like: [dist_mean, dist_ground] of a point-cloud can hold out hypercubes, predict the number of nodes and structure of links there, and create a small subset of lens functions for best generalization performance. Bad prediction performance with good models/mapper params denotes an unpredictable/irregular/unexpected/special part of an object.
- [Self-Dissimilarity](https://en.wikipedia.org/wiki/Self-dissimilarity):
> For many systems characterized as “complex” the patterns exhibited on different scales differ
markedly from one another. For example the biomass distribution in a human body “looks very
different” depending on the scale at which one examines it. Conversely, the patterns at different
scales in “simple” systems (e.g., gases, mountains, crystals) vary little from one scale to another.
Accordingly, the degrees of self-dissimilarity between the patterns of a system at various scales
constitute a complexity “signature” of that system. Here we present a novel quantification of selfdissimilarity.
> This signature can, if desired, incorporate a novel information-theoretic measure of
the distance between probability distributions that we derive here. Whatever distance measure is
chosen, our quantification of self-dissimilarity can be measured for many kinds of real-world data.
This allows comparisons of the complexity signatures of wholly different kinds of systems (e.g.,
systems involving information density in a digital computer vs. species densities in a rain-forest
vs. capital density in an economy, etc.)

With HDBSCAN.

With [Normalized Compression Distance](https://en.wikipedia.org/wiki/Normalized_compression_distance) you can use data compressors as lenses: http://mlwave.github.io/tda/mapper_visualization_ncd.html

Experiment using `[BZ2 Compression ratio, t-SNE 1-D of Normalized Compression^Snappy Distance Matrix]` on binary files, clustering on the projection.

Coarse-Graining

- [Optimal High Level Descriptions of Dynamical Systems](https://arxiv.org/abs/1409.7403)
> To analyze high-dimensional systems, many fields in science and engineering rely on high-level descriptions, sometimes called "macrostates," "coarse-grainings," or "effective theories". Examples of such descriptions include the thermodynamic properties of a large collection of point particles undergoing reversible dynamics, the variables in a macroeconomic model describing the individuals that participate in an economy, and the summary state of a cell composed of a large set of biochemical networks. 
> Often these high-level descriptions are constructed without considering the ultimate reason for needing them in the first place. Here, we formalize and quantify one such purpose: the need to predict observables of interest concerning the high-dimensional system with as high accuracy as possible, while minimizing the computational cost of doing so. The resulting State Space Compression (SSC) framework provides a guide for how to solve for the {optimal} high-level description of a given dynamical system, rather than constructing it based on human intuition alone. 
> In this preliminary report, we introduce SSC, and illustrate it with several information-theoretic quantifications of "accuracy", all with different implications for the optimal compression. We also discuss some other possible applications of SSC beyond the goal of accurate prediction. These include SSC as a measure of the complexity of a dynamical system, and as a way to quantify information flow between the scales of a system.

**Comment by MLWave** [comment]
[Dionysus](http://www.mrzv.org/software/dionysus/) has [Python bindings](http://www.mrzv.org/software/dionysus/python/overview.html#module-dionysus), but I haven't tried it out yet.

**Comment by yuzuhikorunrun** [comment]
I do wonder if you guys are ever planning to enable the feature so that we canselect and download groups of nodes. (for data validation purpose mostly).

Also, for the analysis part, we need to be able to find out the "driver"for each subgroups so that we know what drives separation (maybe some nonparametric tests with p values and scores)....Let me know if that make sense... 

thanks.

**Comment by MLWave** [comment]
> I do wonder if you guys are ever planning to enable the feature so that we canselect and download groups of nodes. (for data validation purpose mostly).

I guess this would require a local web server. A local web server would enable lots of interesting features and is the ideal implementation we are working towards. But no ETA on this, just finishing up first update of the visualization.

> Also, for the analysis part, we need to be able to find out the "driver"for each subgroups so that we know what drives separation (maybe some nonparametric tests with p values and scores)....Let me know if that make sense...

There is 1-2-3 STD rule (feature std diff from inverse_X mean), there is also training whitebox models with the cluster samples labeled as 1, and a random sampling of all the samples labeled as 0.

HDBSCAN has interesting outlier and density statistics (we can use for either entire inverse_X or on the cluster data).

If you want to contribute specifications or code: Consider you have an array of the inverse_X of each cluster, and the complete inverse_X. For every dimension you have a list of natural language description strings.

**Comment by sauln** [comment]
@yuzuhikorunrun, incorporating both of those ideas are definitely on the horizon. I think the Jupyter notebook as @MLWave mentioned would be a great first step. There is a function `data_from_cluster_id` that does want you want, just not in the loop.

If you'd like to work on this, pull requests would be very welcome! Let me know if you need any help.

**Comment by MLWave** [comment]
Working on right-click square selections, this would show the selected cluster ids in a list. Also, drawing graphs could be a function that can be called multiple times (painting multiple complexes on the screen).

So imagine some flow like this:
```
graph_all = map(inverse_X, cover=Cubical(5))

> mouse-select a subset from the graph as subset_ids

subset_X = data_from_subset(subset_ids)

graph_subset = map(subset_X, cover=Cubical(20))

visualize([graph_all, graph_subset])
```

**Comment by MLWave** [comment]
Funny aside: I wonder if with multi-scale mapper, if KeplerMapper is then Turing complete? You can already use KeplerMapper to calculate if a digit is even or uneven (by representing the digit 7 as [1,1,1,1,1,1,1] and setting n_cubes=2). Then the node sizes of the graph are a perfect model for a digit being even or uneven. With multi-scale coverings you could show if a digit is a prime number.

**Comment by sauln** [comment]
> I wonder if with multi-scale mapper, if KeplerMapper is then Turing complete

That would be hilarious :rofl:  Maybe someone will better theoretical CS chops than I try proving this. 

I don't follow how km could show even or odd? Would 6 ([1,1,1,1,1,1]) and 7 ([1,1,1,1,1,1,1]) show up in different clusters?

**Comment by MLWave** [comment]
> I don't follow how km could show even or odd? Would 6 ([1,1,1,1,1,1]) and 7 ([1,1,1,1,1,1,1]) show up in different clusters?

Every number is its own dataset and produces its own graph. 6 ([1,1,1,1,1,1]) with 2 cubes and 0. overlap would produce nodes: [1,1,1] [1,1,1] which would be the same size (and have the same color if the color function looks at cluster member size). 7 ([1,1,1,1,1,1,1]) with the same settings would produce [1,1,1,1] [1,1,1], which graph would look different from even numbers.

You could probably also map numbers within a single dataset. But then you'd need to zero-fill to make the rows of equal size: 6 [1,1,1,1,1,1,0,0,0,0] 7 [1,1,1,1,1,1,1,0,0,0].

All very nerdy though. But if Excel is Turing-complete why not KeplerMapper?

**Comment by MLWave** [comment]
The latest and greatest from Deep Learning is Deep Graph and Deep Manifold Learning. I've always likened mapper as a multi-dimensional convnet on arbitrary data, and I wonder if there is some overlap/potential.

NIPS 2017 tutorial: [Geometric Deep Learning on Graphs and Manifolds](https://nips.cc/Conferences/2017/Schedule?showEvent=8735)
> Michael Bronstein · Joan Bruna · Arthur Szlam · Xavier Bresson · Yann LeCun
> The purpose of the proposed tutorial is to introduce the emerging field of geometric deep learning on graphs and manifolds, overview existing solutions and applications for this class of problems, as well as key difficulties and future research directions.

[Geometric deep learning on graphs and manifolds using mixture model CNNs](https://arxiv.org/abs/1611.08402)
>  Most of deep learning research has so far focused on dealing with 1D, 2D, or 3D Euclidean-structured data such as acoustic signals, images, or videos. Recently, there has been an increasing interest in geometric deep learning, attempting to generalize deep learning methods to non-Euclidean structured data such as graphs and manifolds, with a variety of applications from the domains of network analysis, computational social science, or computer graphics. In this paper, we propose a unified framework allowing to generalize CNN architectures to non-Euclidean domains (graphs and manifolds) and learn local, stationary, and compositional task-specific features. We show that various non-Euclidean CNN methods previously proposed in the literature can be considered as particular instances of our framework. We test the proposed method on standard tasks from the realms of image-, graph- and 3D shape analysis and show that it consistently outperforms previous approaches.

This paper does 3D shape classification, but I can see no comparison to using the mapper approach with Memoli's Gromov-Hausdorff method. I bet, a proper TDA benchmark for 3D shape comparison on current 3D and 2D datasets would already be valuable for this field of research. Going beyond their, rather simplistic, nearest neighbor mapping convolutions could be a major contribution.

**Comment by MLWave** [comment]
Been experimenting with Connectomics and TDA. I have a dataset of a 1000 neurons of C. Elegans, with around 180000 neural activity timeseries, and for the 1 million neuron combinations whether they are connected or not.

Correlation distance matrices cluster many connected neurons together. L2norm lens over the distance matrices creates flares with, what I assume, are different types of neurons.

There is this call-for-research: https://ai-on.org/2017/11/20/discovering-manifold-psychiatric/

> Psychiatric disorders are amongst the most difficult to accurately diagnose and design a treatment plan for. Imaging the structural and functional properties of an individual’s brain is the key to solving this challenge. Current machine learning approaches fail to utilize the information in these brain scans. Broadly, we propose finding a manifold of structural and/or functional brain scans in an embedding space that clusters mental disease states into clinically recognized classes. Specially, we propose a flavour of the adversarial autoencoder to accomplish this task. The goal is wide and ambitious enough to accommodate several other alternatives to solve this challenge.

And I think TDA should be able to help here too. I'll look if I can get reasonable access to their datasets (or existing canonical ones).

**Comment by MLWave** [comment]
Map a 2-D color image with overlapping intervals and different resolutions. Shuffle the pixels inside the interval and reclassify with a pretrained Neural net (ResNet 50 trained on ImageNet): If classification probability goes down: image block is important, if classification probability goes up: image block is not important.

![feature importance](https://i.imgur.com/TVg6vAK.jpg)

We get results similar to LIME and extend "permutation feature importance" to unstructured data.
